src/write_gcm.c

```diff
@@ -597,26 +596,23 @@ static int wg_curl_get_or_post(char *response_buffer,
   return result;
 }
 
-static size_t wg_curl_write_callback(char *ptr, size_t size, size_t nmemb,
+static size_t wg_curl_write_callback(void *ptr, size_t size, size_t nmemb,
                                      void *userdata) {
-  wg_curl_write_ctx_t *ctx = userdata;
-  if (ctx->size == 0) {
+  size_t requested_bytes = size * nmemb;
+  wg_curl_write_ctx_t *ctx = (wg_curl_write_ctx_t *) userdata;
+
+  char *new_data = realloc(ctx->data, ctx->size + requested_bytes + 1);
+  if (new_data == NULL) {
+    ERROR("wg_curl_write_callback: not enough memory (realloc returned NULL)");
     return 0;
```

Comment:
I now realize that @kosak is correct, and we have no indication in the caller that we could not allocate memory for a response (and thus we're retaining the old buffer).
My suggestion would be to at least set `ctx->size` to `-1` and check for that in the caller. We don't need to return the correct size anyway, because `ctx->data` is a NUL-terminated string.